utils/fixed_decimal/src/decimal.rs

```diff
@@ -350,65 +350,106 @@ impl FixedDecimal {
     /// let mut dec = FixedDecimal::from(42);
     /// assert_eq!("42", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(2);
+    /// dec.padded_left(4);
     /// assert_eq!("0042", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(-2);
-    /// assert_eq!("42", dec.to_string());    
+    /// dec.padded_left(3);
+    /// assert_eq!("042", dec.to_string());
     ///
-    /// dec.pad_or_truncate_left(-1);
-    /// assert_eq!("2", dec.to_string());
+    /// dec.padded_left(2);
+    /// assert_eq!("42", dec.to_string());
+    ///
+    /// dec.padded_left(1);
+    /// assert_eq!("42", dec.to_string());
     /// ```
-    pub fn pad_or_truncate_left(&mut self, shift: i16) {
-        self.upper_magnitude = match shift >= 0 {
-            true => self.upper_magnitude.checked_add(shift).unwrap_or(i16::MAX),
-            false => self.upper_magnitude.checked_add(shift).unwrap_or(0),
-        };
-
-        // upper_magnitude must be greater than or equal zero.
-        if self.upper_magnitude < 0 {
-            self.upper_magnitude = 0;
+    pub fn padded_left(&mut self, digits: u16) {
+        let mut magnitude = if digits == 0 { 0 } else if digits > (i16::MAX as u16) + 1 { i16::MAX } else { (digits - 1) as i16 };
+        if magnitude < 0 {
+            magnitude = 0;
         }
-
-        if self.upper_magnitude >= self.magnitude {
-            return;
+        // Do not truncate nonzero digits
+        if magnitude <= self.magnitude {
+            magnitude = self.magnitude;
         }
 
-        // how many digits need to be deleted from the start.
-        let cut = self.magnitude - self.upper_magnitude;
+        self.upper_magnitude = magnitude;
```

Comment:
Nit: add

```rust
        #[cfg(debug_assertions)]
        self.check_invariants();
```

Reply:
^ Please add here, too.

Reply:
Still not here yet.  Please add the `check_invariants` in all three functions (pad left, pad right, truncate left).

Also please fix the clippy warning. :)